### PR TITLE
[Bugfix:TAGrading] Fix Incorrect Stats for Grader Scores

### PR DESCRIPTION
**Pull request by Youssef-Chip**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
Currently, in the grading stats page. the statistics available for each grader (count,average,stddev) are not correctly calculated:
 
image
Notice how the stats for the grader for a specific component differ from the overall stats per components.
Additionally, the stats for the grader per component don't interact with the newly added filters.

### What is the new behavior?
image
The stats for the grader per component are correctly calculated and interact with the filters 
### Other information?
Additionally, this PR reformats some queries related to the statistics for better code practise.

### CODE REVIEW

File: site/app/libraries/database/DatabaseQueries.php [hunk 1]
```diff
@@ -2268,22 +2268,54 @@ public function getAverageComponentScores($g_id, $section_key, $is_team, string
...
+        $null_section_condition = "";
+        $bad_submissions_condition = "";
+        $bad_submission_count = 0;
+        $params = [$gc_id, $g_id, $g_id, $g_id];
         if ($is_team) {
             $u_or_t = "t";
             $users_or_teams = "gradeable_teams";
             $user_or_team_id = "team_id";
         }
+        if ($null_section != 'include') {
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2284]: It is better to use PHP's strict comparison operators in general.

```suggestion
        if ($null_section !== 'include') {
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 2]
```diff
@@ -2268,22 +2268,54 @@ public function getAverageComponentScores($g_id, $section_key, $is_team, string
...
+                    AND ldc.g_id=gc.g_id
+                    AND ldc.submission_days_late > ldc.late_day_exceptions
+                    AND ldc.late_days_change = 0
+        )",
+                [$g_id]
+            );
+            $bad_submission_count = $this->course_db->row()['cnt'];
+
+        // Check if we want to include late (bad) submissions into the average
+        if ($bad_submissions != 'include' && $bad_submission_count > 0) {
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2304]: Same here.

```suggestion
        if ($bad_submissions !== 'include' && $bad_submission_count > 0) {
```

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Review by NiaHeermance — APPROVED** [review]
Looks good. I did a few tests by having multiple graders grade and the grading stats seemed correct.

File: site/app/models/SimpleStat.php [hunk 3]
```diff
@@ -54,6 +54,8 @@ class SimpleStat extends AbstractModel {
     public function __construct(Core $core, $details = []) {
         parent::__construct($core);
         if (isset($details['gc_id'])) {
+            $bad_submissions_cookie = array_key_exists('include_bad_submissions', $_COOKIE) ? $_COOKIE['include_bad_submissions'] : 'omit';
+            $null_section_cookie = array_key_exists('include_null_section', $_COOKIE) ? $_COOKIE['include_null_section'] : 'omit';
```
Comment (williamjallen) [inline site/app/models/SimpleStat.php:58]: This is a great place to use the [null coalescing](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce) operator.

```suggestion
            $bad_submissions_cookie = $_COOKIE['include_bad_submissions'] ?? 'omit';
            $null_section_cookie = $_COOKIE['include_null_section'] ?? 'omit';
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 4]
```diff
@@ -2268,22 +2268,54 @@ public function getAverageComponentScores($g_id, $section_key, $is_team, string
...
+
+        // Check if we want to include late (bad) submissions into the average
+        if ($bad_submissions !== 'include' && $bad_submission_count > 0) {
+            $bad_submissions_condition = "INNER JOIN(
+                SELECT DISTINCT ldc.{$user_or_team_id}
+                FROM late_day_cache AS ldc
+                WHERE ldc.g_id=? AND ( submission_days_late = 0 OR ldc.late_days_change != 0 
+              ) )AS ldc ON ldc.{$user_or_team_id}={$u_or_t}.{$user_or_team_id}";
+            $params[] = $g_id;
+        }
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2311]: Couldn't this entire block of code be added to the giant SQL query below?  We want to minimize the number of SQL queries executed if possible.  This isn't necessarily a requested change that has to be made, just something to think about if it's easy to do.

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/models/SimpleStat.php (hunk 3 above)
Comment (NiaHeermance) [reply site/app/models/SimpleStat.php:58]: I love that operator! A good feature of PHP.

**Review by ziesski — COMMENTED** [review]
Forgot to push the comment, already got merged but I tested it and works. I have not stress tested corner cases tho.

**Pull request merged**